### PR TITLE
Allow dataarrays and preserve dtype in rad2temp

### PR DESCRIPTION
**Pull request by mraspaud**
This PR allow xarray's dataarrays to be passe to radiance to temperature conversions, and also makes these functions preserve dtype.

 - [x] Tests added 
 - [x] Tests passed: Passes ``pytest pyspectral`` 
 - [x] Passes ``flake8 pyspectral`` 
 - [x] Fully documented

### CODE REVIEW

**Review by pnuu — APPROVED** [review]
LGTM.

I love how much you could just delete from the original code :joy:

File: pyspectral/utils.py [hunk 1]
```diff
@@ -580,3 +582,32 @@ def are_instruments_identical(name1, name2):
...
+                return f(*args, **kwargs)
+            import dask.array as da
+            import xarray as xr
+            if isinstance(array, da.Array):
+                return da.map_blocks(f, *args, **kwargs)
+            elif isinstance(array, xr.DataArray):
+                new_array = array.copy()
+                new_args = list(args)
+                new_args[argument_index] = array.data
+                new_data = da.map_blocks(f, *new_args, **kwargs)
```
[the remark this answers is not in the document]
Comment (mraspaud) [reply pyspectral/utils.py:607]: No such requirement that can see or experience.
https://docs.dask.org/en/stable/generated/dask.array.map_blocks.html

File: pyspectral/tests/test_rad_tb_conversions.py [hunk 2]
```diff
@@ -327,33 +320,106 @@ def test_get_bandname(self, download_rsr, load, isfile, exists):
...
+
+    rad = xr.DataArray(da.from_array(np.array([-1, 2, 3]), chunks=2).astype(np.float32), dims=["x"])
+    with warnings.catch_warnings():
+        warnings.simplefilter("error")
+        tbs = xr.map_blocks(radiance2tb, rad, kwargs=dict(wavelength=central_wavelength))
+        assert np.isnan(tbs[0])
+
+    rad = da.from_array(np.array([-1, 2, 3]), chunks=2).astype(np.float32)
+    with warnings.catch_warnings():
+        warnings.simplefilter("error")
```
[the remark this answers is not in the document]
Comment (mraspaud) [reply pyspectral/tests/test_rad_tb_conversions.py:416]: I want to make sure that the code doesn't warn at all, so I'm triggering an error for any warning.
However the only warning the tested code is susceptible to issue is the one about illegal values passed to log.

File: pyspectral/utils.py [hunk 3]
```diff
@@ -580,3 +582,32 @@ def are_instruments_identical(name1, name2):
...
+        def wrapper(*args, **kwargs):
+            array = args[argument_index]
+            chunks = getattr(array, "chunks", None)
+            if chunks is None:
+                return f(*args, **kwargs)
+            import dask.array as da
+            import xarray as xr
+            if isinstance(array, da.Array):
+                return da.map_blocks(f, *args, **kwargs)
+            elif isinstance(array, xr.DataArray):
```
[the remark this answers is not in the document]
Comment (mraspaud) [reply pyspectral/utils.py:603]: I thought about that but considered it to be a generalization I didn't need to do yet. But it's a good idea to put some hint in the docstring, I'll do that now.

[the remark this answers is not in the document]
Comment (mraspaud) [reply pyspectral/utils.py:606]: Done in ec59df7

File: pyspectral/tests/test_rad_tb_conversions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (djhoese) [reply pyspectral/tests/test_rad_tb_conversions.py:416]: Ooohhh ok. Makes sense.

**Review by djhoese — APPROVED** [review]

**Comment by pnuu** [comment]
Closes https://github.com/pytroll/pyspectral/issues/201

**Comment by pnuu** [comment]
Closes https://github.com/pytroll/satpy/issues/2613

**Pull request merged**